### PR TITLE
icub-effort-bridge created

### PR DESCRIPTION
**Pull request by francesco-romano**

### CODE REVIEW

File: human-viz-bridge/CMakeLists.txt [hunk 1]
```diff
@@ -85,12 +85,31 @@ target_include_directories(robot-basepose-publisher PUBLIC ${YARP_INCLUDE_DIRS}
                                               
 list(APPEND conf config/robot-basepose-publisher.ini)
 
+# iCub Effort Bridge
+
+add_executable(icub-effort-bridge src/iCubEffortBridge.cpp)
+
+target_link_libraries(icub-effort-bridge YARP::YARP_OS
```
Comment (francesco-romano) [inline human-viz-bridge/CMakeLists.txt]: I think most of the libraries you are linking to are not needed

File: human-viz-bridge/src/iCubEffortBridge.cpp [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+    iDynTree::VectorDynSize m_robotConfiguration;
+    PolyDriver m_robot;
+    ITorqueControl *robotTorques;
+    vector<EffortPublisher> effortsList;
+
+public:
+    double getPeriod()
+    {
+        // module periodicity (seconds).
+        return 0.1;
```
Comment (francesco-romano) [inline human-viz-bridge/src/iCubEffortBridge.cpp]: put this parametric and add the period in the configuration file

File: human-viz-bridge/src/iCubEffortBridge.cpp [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+    }
+    // Main function. Will be called periodically every getPeriod() seconds
+    bool updateModule()
+    {
+        TickTime currentTime = normalizeSecNSec(yarp::os::Time::now());
+        
+        robotTorques->getTorques(m_robotConfiguration.data());
```
Comment (francesco-romano) [inline human-viz-bridge/src/iCubEffortBridge.cpp]: `m_robotConfiguration` is not the proper name to contain the torques

File: human-viz-bridge/src/iCubEffortBridge.cpp [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+using namespace yarp::os;
+using namespace yarp::sig;
+using namespace yarp::dev;
+
+//---------------------------------------------------------------------------
+//Utility functions for parsing INI file
+static bool parseStringListOption(const Value &option, vector<string> &parsedList);
+//---------------------------------------------------------------------------
+
+struct joint {
```
Comment (francesco-romano) [inline human-viz-bridge/src/iCubEffortBridge.cpp]: Structure with capital letter

File: human-viz-bridge/src/iCubEffortBridge.cpp [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+    {
+        TickTime currentTime = normalizeSecNSec(yarp::os::Time::now());
+        
+        robotTorques->getTorques(m_robotConfiguration.data());
+        
+        for (size_t index = 0; index < effortsList.size(); ++index) {
+            EffortPublisher &effort = effortsList[index];
+            if (!effort.publisher) return false;
+            sensor_msgs_Temperature &effortMsg = effort.publisher->prepare();
+            effort.effortMsg.header.stamp = currentTime;
```
Comment (francesco-romano) [inline human-viz-bridge/src/iCubEffortBridge.cpp]: I still think this is unnecessary and `prepare` is enough

File: human-viz-bridge/src/iCubEffortBridge.cpp [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+        m_robotConfiguration.resize(robot_jointList.size());
+        m_robotConfiguration.zero();
+
+        size_t effortIndex = 0;
+        size_t DoFiterator = 1;
+        string robotJoint = "*";
+        vector<string> jointDegrees = {"_pitch", "_roll", "_yaw"};
+        for (size_t jointIndex = 0; jointIndex < robot_jointList.size(); jointIndex++) {
+            
+            effortsList.resize(effortIndex+1);
```
Comment (francesco-romano) [inline human-viz-bridge/src/iCubEffortBridge.cpp]: `std::vector` has the `push_back` method. I think this is right place to use it.

File: human-viz-bridge/src/iCubEffortBridge.cpp [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+            effortsList[effortIndex].jointEffort.nrDoFs = 0;
+            DoFiterator = 0;
+            
+            if (!(robot_jointList[jointIndex].find(robotJoint)!=string::npos)) {
```
Comment (francesco-romano) [inline human-viz-bridge/src/iCubEffortBridge.cpp]: a `not` `not equal`, i.e. `! (... != ... )` is equivalent to `(... == ...)` which is also clearer

Comment (francesco-romano) [inline human-viz-bridge/src/iCubEffortBridge.cpp]: What is the `*` for?

File: human-viz-bridge/src/iCubEffortBridge.cpp [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+               if(robot_jointList[jointIndex].find(jointDegrees[0])!=string::npos || robot_jointList[jointIndex].find(jointDegrees[1])!=string::npos || robot_jointList[jointIndex].find(jointDegrees[2])!=string::npos ) 
+               {
+                   robotJoint = robot_jointList[jointIndex];
+                   size_t found = robotJoint.find("_");
```
Comment (francesco-romano) [inline human-viz-bridge/src/iCubEffortBridge.cpp]: What are you trying to do with this line and the next 7?

File: human-viz-bridge/src/iCubEffortBridge.cpp [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+                   size_t erasePos = robotJoint.find("_",(found+1));
+                   cout << robotJoint << endl;
+                   if (erasePos < 100) {
+                       effortsList[effortIndex].jointEffort.jointName = robotJoint.erase(erasePos);
+                   } else {
+                       effortsList[effortIndex].jointEffort.jointName = robotJoint.erase(found);
+                   }
+                   effortsList[effortIndex].linkEffort = robotModel.getLinkName(traversal.getChildLinkIndexFromJointIndex(robotModel, robotModel.getJointIndex(robot_jointList[jointIndex])));
+                   bool sameJoint = robot_jointList[(jointIndex+(DoFiterator))].find(effortsList[effortIndex].jointEffort.jointName)!=string::npos; 
+                   while(sameJoint){
```
Comment (francesco-romano) [inline human-viz-bridge/src/iCubEffortBridge.cpp]: I think this is extremely complicated! 
Furthermore, what happen if somebody uses a list like
`torso_yaw, r_elbow, torso_roll` ?

File: human-viz-bridge/src/iCubEffortBridge.cpp [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+            return false;
+        }
+        
+        
+        return true;
+    }
+    // Close function, to perform cleanup.
+    bool close()
+    {
+        iCubEffortDataPort.close();
```
Comment (francesco-romano) [inline human-viz-bridge/src/iCubEffortBridge.cpp]: The connection to the robot should also be closed

File: human-viz-bridge/src/iCubEffortBridge.cpp [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+        }
+        effortsList.clear();
+        return true;
+    }
+    
+    bool getRobotTorquesInterface(const yarp::os::Searchable& config,
+                                  const std::vector<std::string>& robot_jointList,
+                                  yarp::dev::ITorqueControl *& torques)
+    {
+        if (!m_robotConfigured) {
```
Comment (francesco-romano) [inline human-viz-bridge/src/iCubEffortBridge.cpp]: you call this function only once. This variable is not needed

File: human-viz-bridge/src/iCubEffortBridge.cpp (hunk 7 above)
Comment (MartaLorenzini) [inline human-viz-bridge/src/iCubEffortBridge.cpp]: It is only a default value for the first iteration, i needed something that for sure is not present in the joint name

File: human-viz-bridge/src/iCubEffortBridge.cpp (hunk 8 above)
Comment (MartaLorenzini) [inline human-viz-bridge/src/iCubEffortBridge.cpp]: To remove the suffix of the joints whose names contain "_roll", "_pitch" or "_yaw"

File: human-viz-bridge/src/iCubEffortBridge.cpp (hunk 9 above)
Comment (MartaLorenzini) [inline human-viz-bridge/src/iCubEffortBridge.cpp]: It would a bad person. By the way I will think about a less complicated way to do that :D

**Pull request closed without merging**